airbyte-ci: Add pypi publishing logic

flash1293 opened the pull request:
Add pypi publish functionality as a standalone command and as step within connector publish:

Standalone command is in a group `poetry`:

This is how to publish a poetry project like airbyte lib:
```
airbyte-ci poetry --package-path=airbyte-lib publish --pypi-token="api-token" --test-pypi
```

It can also publish legacy projects (this is the functionality used by the connector publish pipeline):
```
airbyte-ci poetry --package-path=airbyte-integrations/connectors/source-apify-dataset publish --publish-name=airbyte-source-apify-dataset --publish-version="2.1.2" --pypi-token="api-token" --test-pypi
```


During connector publish, it checks whether the pypi remote registry is enabled - if yes, it checks whether the current version is published already. If no, it will run the same logic as the `poetry publish` command from above, with the package name set to `airbyte-{connector-name}` and version set to the docker image tag.

Internally, the package name and version provided to the step will override whatever is defined in the setup.py or pyproject.toml file of the package.


# Open questions:
* How to pass the pypi api token to the connector publish step? Currently, it's just reading from os.environ deep in the code, a cleaner approach would probably be nice, but putting this very specific thing on the top level of airbyte-ci options also feels wrong. Happy for expert opinions here
* Is the basic setup any good or should it be structured differently?
* What image should be used to perform the publish? I just went with an image that has poetry pre-installed, but maybe we have some more defined opinions on it. Ideally I would like to not block on this for the initial release of this feature to unblock the functionality first as this is a critical part
* Integration of passing the pypi api token via CI - I need someone with admin rights to configure this on the repo I think
* Testing - what are the best strategies to test this? We can of course mock stuff away but it would be nice to stay as close as possible to the "real" thing. One thing that's difficult with this is that we can't publish the same version multiple times, pypi doesn't allow this


Test connector publish pipeline via https://github.com/airbytehq/airbyte/actions/workflows/publish_connectors.yml

bnchrch commented:
Ill have to give the code a review later today but let me answer those open questions!

## Open Questions

> How to pass the pypi api token to the connector publish step? Currently, it's just reading from os.environ deep in the code, 
a cleaner approach would probably be nice, but putting this very specific thing on the top level of airbyte-ci options also feels wrong. Happy for expert opinions here

Reading the [config docs](https://python-poetry.org/docs/configuration/#pypi-tokenname) for poetry it seems as though they have env variable options for their config params out of the box.

To use these env vars we would have to
1. set `POETRY_PYPI_TOKEN_<NAME>` in the host runner
2. tell dagger to pass this through to the container thats running as a secret (heres an [example](https://github.com/airbytehq/airbyte/blob/f90d8dbccb768e62cbddce178d6ec3af8d4ac8ea/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/docker.py#L79) of how this is done)

We could either do this explicitly for the poetry publish pipeline or add a new generic option to the airbyte-ci that is meant to pass arbitrary env vars to the running containers.

I dont want to blow up your scope though and would think that explicitly defining this secret would be fine

> Is the basic setup any good or should it be structured differently?

Given that this is connector specific my only to high level structural asks is
1. `connector poetry publish` is a standalone step (`ConnectorPypiPublishStep`)
2. That step is used in the publish pipeline and is added/enabled to the `STEP_TREE` based on the metadata file option
3. We can "only" run publish by running `airbyte-ci conenctor publish` and using `-x` to turn off unwanted steps
4. If we implement a pypi only publish step it reuses `ConnectorPypiPublishStep`

> What image should be used to perform the publish? I just went with an image that has poetry pre-installed, but maybe we have some more defined opinions on it. Ideally I would like to not block on this for the initial release of this feature to unblock the functionality first as this is a critical part

I think whatever one builds the connector makes sense. I believe we bundle poetry with the builder step. If not we can.

> Integration of passing the pypi api token via CI - I need someone with admin rights to configure this on the repo I think
Yes thats correct. But I think you should recieve admin right! See if AJ can grant that :)

>Testing - what are the best strategies to test this? We can of course mock stuff away but it would be nice to stay as close as possible to the "real" thing. One thing that's difficult with this is that we can't publish the same version multiple times, pypi doesn't allow this

I know pypi has a test repository that behaves identical to the main repo. Can we simply use that? + prerelease tags based on gitsha to allow us to repeatedly publish?

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/pipeline.py:
```diff
@@ -273,6 +291,22 @@ def create_connector_report(results: List[StepResult]) -> ConnectorReport:
                 metadata_upload_results = await metadata_upload_step.run()
                 results.append(metadata_upload_results)
 
+            # Try to convert the context to a PyPIPublishContext. If it returns None, it means we don't need to publish to pypi.
```
Can you add the unit tests in this [module](https://github.com/airbytehq/airbyte/blob/main/airbyte-ci/connectors/pipelines/tests/test_publish.py#L16) to check the behavior following your addition?

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/commands.py:
```diff
@@ -0,0 +1,102 @@
+    is_flag=True,
+    default=False,
+)
+@click.option(
+    "--publish-name",
+    help="The name of the package to publish. If not set, the name will be inferred from the pyproject.toml file of the package.",
+    type=click.STRING,
+)
+@click.option(
+    "--publish-version",
+    help="The version of the package to publish. If not set, the version will be inferred from the pyproject.toml file of the package.",
+    type=click.STRING,
+)
```
What are the use cases for these options? Do we ever want to override a package name / version?

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/pipeline.py:
```diff
@@ -0,0 +1,209 @@
+        if "connectorBuildOptions" in current_metadata and "baseImage" in current_metadata["connectorBuildOptions"]:
+            build_docker_image = current_metadata["connectorBuildOptions"]["baseImage"]
+        else:
+            build_docker_image = "mwalbeck/python-poetry"
```
Can we use our own base image? It has poetry.

alafanechere commented:
@flash1293 @bnchrch Just adding my two cents as I think there's some confusion in:
> Given that this is connector specific my only to high level structural asks is
connector poetry publish is a standalone step (ConnectorPypiPublishStep)
That step is used in the publish pipeline and is added/enabled to the STEP_TREE based on the metadata file option
We can "only" run publish by running airbyte-ci conenctor publish and using -x to turn off unwanted steps
If we implement a pypi only publish step it reuses ConnectorPypiPublishStep

* @flash1293 declared a `PublishToPyPi` step meant to be generic enough to be re-used for non connector work (like releasing airbyte-lib, metadata-lib or the to PyPi.
* The `poetry` command group is meant to be used under `airbyte-ci` to publish our internal packages (airbyte-lib, metadata-lib etc.). with `airbyte-ci poetry publish`. We'll move `test` under `poetry` to test these packages.
* The connector publish flow is also touched in this PR. The STEP TREE logic is not used in the publish pipeline at the moment, and I'm not sure we should expose options to customize the flow here. Skipping pypi publishing according to metadata content LGTM.

bnchrch commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/pipeline.py:
```diff
@@ -273,6 +291,22 @@ def create_connector_report(results: List[StepResult]) -> ConnectorReport:
                 results.append(metadata_upload_results)
 
+            # Try to convert the context to a PyPIPublishContext. If it returns None, it means we don't need to publish to pypi.
+            pypi_context = await PyPIPublishContext.from_connector_context(context)
```
💅 Lets ensure we move this logic to a named function to keep `run_connector_publish_pipeline` readable and focused on coordination

bnchrch commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/commands.py:
```diff
@@ -0,0 +1,102 @@
+)
+@pass_pipeline_context
+@click.pass_context
+async def publish(
```
😍 Loving the (relative) simplicity

bnchrch commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/pipeline.py:
```diff
@@ -0,0 +1,209 @@
+
+        if is_pip_package:
+            # legacy publish logic
+            pypi_username = self.context.dagger_client.set_secret("pypi_username", "__token__")
```
💡Given that we are following a simple convention of
1. Choose docker image
2. Set secrets
3. Mount paths
4. Run command

You may want to look at `SimpleDockerStep`. Heres an example of it in use

https://github.com/airbytehq/airbyte/blob/99488ddcc50fd16b8753019864e80b37b947a76d/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/metadata/pipeline.py#L24

flash1293 replied:
I tried to convert it this way but due to the branching and some extras like writing certain files it got messy and IMHO it's clearer when cleaned up like @alafanechere described above - I'm still working through some comments, let me finish that and have a look at the final results, happy to do another pass.

bnchrch reviewed (commented):
Not much to add ontop of Augustins review, just a few small ideas that may help clarity.

Thanks for diving in Joe!

flash1293 commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/pipeline.py:
```diff
@@ -52,6 +54,22 @@ async def _run(self) -> StepResult:
             return StepResult(self, status=StepStatus.SUCCESS, stdout=f"No manifest found for {self.context.docker_image}.")
 
 
+class CheckPypiPackageExists(Step):
```
I didn't plan to - when running it via the command IMHO the caller is responsible for this and giving them the exact error from pypi is the best behavior.

flash1293 commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/pipeline.py:
```diff
@@ -273,6 +291,22 @@ def create_connector_report(results: List[StepResult]) -> ConnectorReport:
                 metadata_upload_results = await metadata_upload_step.run()
                 results.append(metadata_upload_results)
 
+            # Try to convert the context to a PyPIPublishContext. If it returns None, it means we don't need to publish to pypi.
```
Currently the behavior is like this:
* If pypi is not enabled, the steps are not even shown
* If pypi is enabled but it's published already, the steps are shown as skipped

My reasoning was that it might be weird / unexpected to even show pypi-related stuff on connector reports that have nothing to do with it. Basically assuming the following semantic meaning of skip: "Normally this step would be performed, but an edge case prevent us from doing so - you might look into that".

No strong opinion on this though, we can also always show them as "skipped" if you think it makes sense.

alafanechere replied:
Agreed, it makes no sense to show skipped pypi steps for a java connector.

flash1293 commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/pipeline.py:
```diff
@@ -0,0 +1,209 @@
+            # legacy publish logic
+            pypi_username = self.context.dagger_client.set_secret("pypi_username", "__token__")
+            pypi_password = self.context.dagger_client.set_secret("pypi_password", f"pypi-{self.context.pypi_token}")
+            metadata = {
+                "name": self.context.package_name,
+                "version": self.context.version,
+                "author": "Airbyte",
+                "author_email": "contact@airbyte.io",
+            }
```
You mean author and author_email? If yes I did that on purpose to enforce consistency. Added a comment for this.

flash1293 commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/pipeline.py:
```diff
@@ -0,0 +1,209 @@
+                .with_directory("package", dir_to_publish)
+                .with_workdir("package")
+                .with_new_file("pyproject.toml", contents=tomli_w.dumps(contents))
+                .with_exec(["poetry", "config", "repositories.testpypi", "https://test.pypi.org/legacy/"])
```
It doesn't seem to hurt to do so

flash1293 commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/pipeline.py:
```diff
@@ -0,0 +1,209 @@
+        If pypi is not enabled, this will return None.
+        """
+
+        current_metadata = yaml.safe_load(
```
This doesn't work because `connector_context.connector.metadata` is what got published the last time, not the current state in the repo, so it wouldn't publish on the first merge that sets `remoteRegistries.pypi.enabled` to true.

flash1293 replied:
Good point, I could have sworn I ran into problems with this, that's why I changed it, but maybe I confused myself. I will look into this.

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/commands.py:
```diff
@@ -0,0 +1,102 @@
+    name="poetry",
+    help="Commands related to running poetry commands.",
+)
+@click.option(
```
> it might be beneficial to directly publish in some cases (especially in early phases).

In these cases would you publish from a local airbyte-ci execution?
If yes, I would suggest adding a user prompt instead of logging a warning.

flash1293 replied:
Yes, exactly. Makes sense to me, switching to a prompt

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/commands.py:
```diff
@@ -0,0 +1,82 @@
+        version=publish_version,
+    )
+
+    if context.package_path.startswith(CONNECTOR_PATH_PREFIX):
```
I'd prefer the check "is it a connector" being: does this package have a `metadata.yaml` file. 
It will make this command less coupled to the repository structure.
We're considering allowing airbyte-ci to run on external repo whose structure can be different from ours.

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/context.py:
```diff
@@ -0,0 +1,95 @@
+        )
+
+    @staticmethod
+    async def from_publish_connector_context(connector_context: PublishConnectorContext) -> Optional["PyPIPublishContext"]:
```
```suggestion
    from future import __annotations__
    ...
    async def from_publish_connector_context(connector_context: PublishConnectorContext) -> Optional[PyPIPublishContext]:
```

flash1293 replied:
Not sure I understand the change - could you explain?

alafanechere replied:
The `from future import __annotations__` allows you to add type hints for a local class that has not yet been evaluated. Quoting the class name works otherwise, but I find it cleaner to not quote it.

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/context.py:
```diff
@@ -0,0 +1,95 @@
+            ci_gcs_credentials=ci_gcs_credentials,
+        )
+
+    @staticmethod
+    async def from_publish_connector_context(connector_context: PublishConnectorContext) -> Optional["PyPIPublishContext"]:
```
```suggestion
    @classmethod
    async def from_publish_connector_context(cls: PyPIPublishContext, connector_context: PublishConnectorContext) -> Optional["PyPIPublishContext"]:
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/context.py:
```diff
@@ -0,0 +1,95 @@
+        version = current_metadata["dockerImageTag"]
+        if connector_context.pre_release:
+            # use current date as pre-release version
+            rc_tag = datetime.now().strftime("%Y%m%d%H%M")
```
```suggestion
            release_candidate_tag = datetime.now().strftime("%Y%m%d%H%M")
```
Can we somehow use the [same tagging logic as for docker pre-release images](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/context.py#L93) using commit sha? Would be cool to use a common function to tag pre-release for docker images and pypi packages. And I kinda like the fact that tagging / version is handled at the Context level.

flash1293 replied:
We could do the git commit hash thing when using local version identifiers: https://packaging.python.org/en/latest/specifications/version-specifiers/#local-version-identifiers

But they need to be prefixed with a `+`, so it won't be 100% the same. Still a bit better than the current local logic, I will adjust.

flash1293 replied:
Just tested and unfortunately it doesn't work - test.pypi is rejecting local versions:
> 400 '0.2.1+391b68cd52' is an invalid value for Version. Error:   
         Can't use PEP 440 local versions. See                                  
         https://packaging.python.org/specifications/core-metadata for more     
         information.

Public version identifiers need to be numeric, I will revert back to the old logic.

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/commands.py:
```diff
@@ -0,0 +1,82 @@
+    "--registry-url",
+    help="Which registry to publish to. If not set, the default pypi is used. For test pypi, use https://test.pypi.org/legacy/",
+    type=click.STRING,
+    default="https://pypi.org/simple",
```
```suggestion
    default=DEFAULT_PYTHON_REGISTRY_URL,
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/commands.py:
```diff
@@ -0,0 +1,82 @@
+@click.option(
+    "--publish-version",
+    help="The version of the package to publish. If not set, the version will be inferred from the pyproject.toml file of the package.",
+    type=click.STRING,
```
nit: shall we add a callback validating the version follows semver?

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/commands.py:
```diff
@@ -0,0 +1,82 @@
+CONNECTOR_PATH_PREFIX = "airbyte-integrations/connectors"
+
+
+@click.command(cls=DaggerPipelineCommand, name="publish", help="Publish a Python package to PyPI.")
```
```suggestion
@click.command(cls=DaggerPipelineCommand, name="publish", help="Publish a Python package to a registry.")
```

Even if your implementation is not supporting private registries  I believe we can consider it's agnostic of the use of PyPi as a registry. PyPi is just the default.

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+    NONE = auto()
+
+
+class PublishToPyPI(Step):
```
```suggestion
class PublishToPythonRegistry(Step):
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+        pyproject_toml = dir_to_publish.file(PYPROJECT_TOML_FILE_PATH)
+        pyproject_toml_content = await pyproject_toml.contents()
+        contents = tomli.loads(pyproject_toml_content)
+        if (
+            "tool" not in contents
+            or "poetry" not in contents["tool"]
+            or "name" not in contents["tool"]["poetry"]
+            or "version" not in contents["tool"]["poetry"]
+        ):
+            return None
+
+        return (contents["tool"]["poetry"]["name"], contents["tool"]["poetry"]["version"])
```
nit: 
```suggestion
       try:
          return (contents["tool"]["poetry"]["name"], contents["tool"]["poetry"]["version"])
       except KeyError:
         return None
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+    def _get_base_container(self) -> Container:
+        return with_poetry(self.context)
+
+    async def _get_package_metadata_from_pyproject_toml(self, dir_to_publish: Directory) -> Optional[Tuple[str, str]]:
```
```suggestion
    async def _get_package_metadata_from_pyproject_toml(self, dir_to_publish: Directory) -> Optional[PackageMetadata]:
```

nit: Building a light dataclass with `name` and `version` attribute could boost readability instead of returning a tuple

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+            return PackageType.NONE
+
+    async def _run(self) -> StepResult:
+        dir_to_publish = await self.context.get_repo_dir(self.context.package_path)
```
nit
```suggestion
        package_dir_to_publish = await self.context.get_repo_dir(self.context.package_path)
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+
+        return (contents["tool"]["poetry"]["name"], contents["tool"]["poetry"]["version"])
+
+    async def _get_package_type(self, dir_to_publish: Directory) -> PackageType:
+        files = await dir_to_publish.entries()
+        has_pyproject_toml = PYPROJECT_TOML_FILE_PATH in files
+        has_setup_py = SETUP_PY_FILE_PATH in files
+        if has_pyproject_toml:
+            return PackageType.POETRY
+        elif has_setup_py:
+            return PackageType.PIP
+        else:
+            return PackageType.NONE
```
```suggestion
    async def _get_package_type(self, dir_to_publish: Directory) -> Optional[PackageType]:
        files = await dir_to_publish.entries()
        has_pyproject_toml = PYPROJECT_TOML_FILE_PATH in files
        has_setup_py = SETUP_PY_FILE_PATH in files
        if has_pyproject_toml:
            return PackageType.POETRY
        elif has_setup_py:
            return PackageType.PIP
        else:
            return None
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+        dir_to_publish = await self.context.get_repo_dir(self.context.package_path)
+        package_type = await self._get_package_type(dir_to_publish)
+
+        if package_type == PackageType.NONE:
```
```suggestion
        if not package_type:
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+        # Try to infer package name and version from the pyproject.toml file. If it is not present, we need to have the package name and version set
+        # Setup.py packages need to set package name and version as parameter
+        if not self.context.package_name or not self.context.version:
+            if not package_type == PackageType.POETRY:
```
```suggestion
            if package_type is not PackageType.POETRY:
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+                    "Connector does not have a pyproject.toml file which specifies package name and version and they are not set otherwise, skipping."
+                )
+
+            self.context.package_name = package_metadata[0]
+            self.context.version = package_metadata[1]
```
```suggestion
            self.context.package_metadata : PackageMetadata = package_metadata
```
nit: Just suggesting bundling package metadata inside a new dataclass.

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+
+        self.logger.info(f"Uploading package {self.context.package_name} version {self.context.version} to {self.context.registry}...")
+
+        if package_type == PackageType.PIP:
```
```suggestion
        if package_type is PackageType.PIP:
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
```diff
@@ -0,0 +1,153 @@
+        else:
+            return PackageType.NONE
+
+    async def _run(self) -> StepResult:
```
Readability improvement suggestion:
Would it be possible to slightly rework this method so that when reading it we don't have to follow the `if poetry` or `if pip`? The switch / fork could happen in sub methods.

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/utils.py:
```diff
@@ -0,0 +1,18 @@
+    :param test_pypi: Set to True to check on Test PyPI, False for regular PyPI.
+    :return: True if the package is found with the specified version, False otherwise.
+    """
+    url = f"{base_url}{package_name}/{version}/json"
```
```suggestion
    url = f"{base_url}/{package_name}/{version}/json"
```
?

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/consts.py:
```diff
@@ -60,6 +60,8 @@
 POETRY_CACHE_PATH = "/root/.cache/pypoetry"
 STORAGE_DRIVER = "fuse-overlayfs"
 TAILSCALE_AUTH_KEY = os.getenv("TAILSCALE_AUTH_KEY")
+PYPROJECT_TOML_FILE_PATH = "pyproject.toml"
```
It's already defined on line 11

alafanechere commented on airbyte-ci/connectors/pipelines/tests/test_poetry/test_poetry_publish.py:
```diff
@@ -0,0 +1,92 @@
+        ),
+    ],
+)
+async def test_run_poetry_publish(context: PyPIPublishContext, package_path: str, package_name: str, expected_asset: str):
```
Would it be worth adding the same kind of integration test for "non poetry" package?

flash1293 replied:
apify-dataset is a `setup.py`-style package and airbyte-lib is a poetry-style package, so both branches are tested via the parameterization.

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py:
I'm not sure this module should be stored under the `airbyte_ci/poetry` package as its also used by connectors. Can we find a more common place? Maybe a new module under  [`pipelines/airbyte_ci/steps`](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps)

flash1293 commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/commands.py:
```diff
@@ -0,0 +1,34 @@
+@click_merge_args_into_context_obj
+@pass_pipeline_context
+@click_ignore_unused_kwargs
+async def poetry(pipeline_context: ClickPipelineContext) -> None:
```
I meant the folder structure mostly - one file for the group, then a folder for each sub-command (just publish for now).

As there isn't really a use case for multiple package paths (as the step can be instrumented from other pipelines) and it's increasing the scope, I would like to split that out for now..

flash1293 commented:
@alafanechere 

I addressed your comments, it's ready for another look.

During testing I think I noticed a weird error with the caching. To test the connector publish, I added the remoteRegistries section for pokeapi and ran the workflow
* On the first try it skipped over the pypi section, even though it was enabled in the metadata: https://github.com/airbytehq/airbyte/actions/runs/7594166736/job/20685401062
* Then, I added a debug statement to the airbyte-ci code and ran again. This time, it correctly recognized that pypi remote publishing is enabled for the connector: https://github.com/airbytehq/airbyte/actions/runs/7594192174

The only thing I changed was the airbyte-ci code, because of this I think there is something cached that shouldn't be cached. It's possible this is only an issue during development because it's running airbyte-ci in dev mode, but I wanted to highlight it - could you take a look at this please to make sure it will work correctly in production?

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/python_registry/context.py:
```diff
@@ -0,0 +1,105 @@
+        version = current_metadata["dockerImageTag"]
+        if connector_context.pre_release:
+            # use current date as pre-release version
+            release_candidate_tag = datetime.now().strftime("%Y%m%d%H%M")
```
@flash1293 can you add a comment here explaining why using the same tagging strategy as docker tags does not work? 🙏

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/python_registry/pipeline.py:
```diff
@@ -0,0 +1,165 @@
+
+class PublishToPythonRegistry(Step):
+    context: PythonRegistryPublishContext
+    title = "Publish package to PyPI"
```
```suggestion
    title = "Publish package to a python registryre"
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/python_registry/pipeline.py:
```diff
@@ -0,0 +1,165 @@
+            return await self._poetry_publish(package_dir_to_publish)
+
+    async def _poetry_publish(self, package_dir_to_publish: Directory) -> StepResult:
+        pypi_token = self.context.dagger_client.set_secret("pypi_token", f"pypi-{self.context.pypi_token}")
```
```suggestion
        python_registy_token = self.context.dagger_client.set_secret("python_registy_token", f"python_registy_token-{self.context.python_registy_token}")
```

Is the `pypi-` prefix required?

flash1293 replied:
we could theoretically move it to the secret as well, I'll remove it from here.

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/python_registry/pipeline.py:
```diff
@@ -0,0 +1,165 @@
+        contents["tool"]["poetry"]["authors"] = ["Airbyte <contact@airbyte.io>"]
+        poetry_publish = (
+            self._get_base_container()
+            .with_secret_variable("PYPI_TOKEN", pypi_token)
```
```suggestion
            .with_secret_variable("PYTHON_REGISTRY_TOKEN", python_registry_token)
```

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/python_registry/pipeline.py:
```diff
@@ -0,0 +1,165 @@
+
+    async def _pip_publish(self, package_dir_to_publish: Directory) -> StepResult:
+        files = await package_dir_to_publish.entries()
+        pypi_username = self.context.dagger_client.set_secret("pypi_username", "__token__")
+        pypi_password = self.context.dagger_client.set_secret("pypi_password", f"pypi-{self.context.pypi_token}")
```
Can we replace all `pypi` occurrence by `python_registry`?

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/python_registry/pipeline.py:
Just for module path consistency:
Could we:
* Only have steps declaration under `airbyte_ci/steps`? This module would be renamed to `airbyte_ci/steps/python_registry.py`
* The `context.py` could go under `models/contexts/python_registry_publish.py`

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/steps/python_registry/utils.py:
```diff
@@ -0,0 +1,29 @@
+from pipelines.airbyte_ci.steps.python_registry.context import PythonPackageMetadata
+
+
+def is_package_published(package_metadata: PythonPackageMetadata, registry_url: str) -> bool:
```
To boost re-use of this function could be split the args and not use `PythonPackageMetadata` here?
The signature could be `name, version, registry_url`. 
You can store this module under `airbyte-ci/connectors/pipelines/pipelines/helpers/pip.py`

Wdyt?

alafanechere commented on airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/pipeline.py:
```diff
@@ -52,6 +54,26 @@ async def _run(self) -> StepResult:
             return StepResult(self, status=StepStatus.SUCCESS, stdout=f"No manifest found for {self.context.docker_image}.")
 
 
+class CheckPypiPackageDoesNotExist(Step):
```
I'm wondering if this Step might be re-used  and would be better placed in the shared module

flash1293 replied:
I actually like it being localized for now - those "generic" places can get crowded quickly. What about leaving it in there for now and moving it out once we need it?

alafanechere reviewed (approved):
Nothing blocking but additional feedback following your latest changes.
I'm wondering to what extent we could remove `pypi` occurences from the code and replace them by `python_registry`. I'd love to understand which logic is pypi specific. If we have some pypi specific logic I'd appreciate clearly isolating them so that supporting private / other python registry is easy.
I'll check your caching issue now.

alafanechere commented:
> @alafanechere
> 
> I addressed your comments, it's ready for another look.
> 
> During testing I think I noticed a weird error with the caching. To test the connector publish, I added the remoteRegistries section for pokeapi and ran the workflow
> 
> * On the first try it skipped over the pypi section, even though it was enabled in the metadata: https://github.com/airbytehq/airbyte/actions/runs/7594166736/job/20685401062
> * Then, I added a debug statement to the airbyte-ci code and ran again. This time, it correctly recognized that pypi remote publishing is enabled for the connector: https://github.com/airbytehq/airbyte/actions/runs/7594192174
> 
> The only thing I changed was the airbyte-ci code, because of this I think there is something cached that shouldn't be cached. It's possible this is only an issue during development because it's running airbyte-ci in dev mode, but I wanted to highlight it - could you take a look at this please to make sure it will work correctly in production?

I'm pretty sure this problems comes from an issue on this [switch](https://github.com/airbytehq/airbyte/blob/master/.github/actions/run-dagger-pipeline/action.yml#L107). 
On your first run the airbyte-ci binary built on master was used, explaining why your new logic was not called.
On the second run airbyte-ci is install from the branch code. 
I believe the "install airbyte-ci dev" works only if you modified airbyte-ci code on your head commit, not on the branch.

alafanechere commented:
@flash1293 I successfully tested a source-pokeapi pre-release with both Pypi [disabled](https://github.com/airbytehq/airbyte/actions/runs/7627051059) and [enabled](https://github.com/airbytehq/airbyte/actions/runs/7627238247/job/20775458399).
